core/src/test/scala/unit/kafka/server/TxnOffsetCommitRequestTest.scala

```diff
@@ -0,0 +1,184 @@
+        generationId = generationId,
+        producerId = producerIdAndEpoch.producerId,
+        producerEpoch = producerIdAndEpoch.epoch,
+        transactionalId = transactionalId,
+        topic = topic,
+        partition = partition,
+        offset = offset,
+        expectedError = Errors.NONE,
+        version = version
+      )
+
+      // Send writeTxnMarker to force transaction from pending to completed
+      // hence we can retrieve offsets from OFFSET_FETCH request
+      val writableTxnMarkersResult = writeTxnMarkers(
+        Collections.singletonList(new WritableTxnMarker()
+          .setTopics(Collections.singletonList(new WritableTxnMarkerTopic()
+            .setName(Topic.GROUP_METADATA_TOPIC_NAME)
+            .setPartitionIndexes(Collections.singletonList(0))))
+          .setProducerId(producerIdAndEpoch.producerId)
+          .setProducerEpoch(producerIdAndEpoch.epoch)
+          .setCoordinatorEpoch(0.toShort)
+          .setTransactionResult(true))
+      )
+      assertEquals(1, writableTxnMarkersResult.size())
+      val writableTxnMarkerTopic = writableTxnMarkersResult.asScala.head.topics().asScala.head
+      assertEquals(Topic.GROUP_METADATA_TOPIC_NAME, writableTxnMarkerTopic.name())
+      assertEquals(1, writableTxnMarkerTopic.partitions().size())
+      assertEquals(Errors.NONE.code(), writableTxnMarkerTopic.partitions().asScala.head.errorCode())
+
+      val fetchOffsetsResp = fetchOffsets(
+        groups = Map(groupId -> List(new TopicPartition(topic, partition))),
+        requireStable = true,
+        version = ApiKeys.OFFSET_FETCH.latestVersion()
+      )
+      val groupIdRecord = fetchOffsetsResp.find(_.groupId() == groupId).head
+      val topicRecord = groupIdRecord.topics().asScala.find(_.name() == topic).head
+      val partitionRecord = topicRecord.partitions().asScala.find(_.partitionIndex() == partition).head
+      assertEquals(offset, partitionRecord.committedOffset())
+    }
+
+    for (version <- 0 to ApiKeys.TXN_OFFSET_COMMIT.latestVersion(isUnstableApiEnabled)) {
+      // Verify that the TXN_OFFSET_COMMIT request is processed correctly when member id is UNKNOWN_MEMBER_ID
+      // and generation id is UNKNOWN_GENERATION_ID under all api versions
+      verifyTxnCommitAndFetch(
+        groupId = groupId,
+        memberId = JoinGroupRequest.UNKNOWN_MEMBER_ID,
+        generationId = JoinGroupRequest.UNKNOWN_GENERATION_ID,
+        offset = 100 + version,
+        version = version.toShort
+      )
+
+      // Verify that the TXN_OFFSET_COMMIT request is processed correctly when the member ID
+      // and generation ID are known. This validation starts from version 3, as the member ID
+      // must not be empty from version 3 onwards.
+      if (version >= 3) {
+        verifyTxnCommitAndFetch(
+          groupId = groupId,
+          memberId = memberId,
+          generationId = memberEpoch,
+          offset = 200 + version,
+          version = version.toShort
+        )
+      }
```

Comment:
Would it be possible to add a negative cases too? We could add them in this test or in another one.

Reply:
Perhaps we can put the negative case to [TxnOffsetCommitRequestTest.java](https://github.com/apache/kafka/blob/trunk/clients/src/test/java/org/apache/kafka/common/requests/TxnOffsetCommitRequestTest.java) ? I thought the TxnOffsetCommitRequestTest should take care the scenario that GroupMetadata is defined while version < 3.
Let me know if you have any concern. Thanks.

Reply:
By negative cases, I meant a request with invalid memberId and/or generationId to ensure that the request is rejected with the correct error. Is it worth it?